File: data-intelligence-system/labeling/mask_codec.py
"""COCO run-length encoding for segmentation masks.

WHY THIS EXISTS RATHER THAN pycocotools
---------------------------------------
pycocotools is the reference implementation, but it is a C extension: adding it
would put a compiler back in the runtime image we just spent effort slimming,
and it pulls its own numpy pin. The format itself is ~60 lines, so it is
implemented here and tested against the cases that actually break naive
encoders (see tests/test_mask_codec.py).

THE FORMAT
----------
COCO's "compressed RLE" is defined by what pycocotools does, not by a spec:

  * Runs are counted in COLUMN-MAJOR (Fortran) order, which is why a mask that
    round-trips under row-major looks transposed in every COCO tool.
  * The sequence always starts with a run of BACKGROUND, which may be length 0
    when the top-left pixel is foreground.
  * Each run length is split into 6-bit chunks, low bits first; bit 0x20 marks
    "more chunks follow"; each chunk is offset by 48 into printable ASCII.
  * From the third run onward the DELTA against the run two positions earlier
    is encoded instead of the absolute length. Adjacent columns of a real mask
    are strongly correlated, so those deltas are small and cost fewer chunks.
    This is where the compression comes from, and it is the part naive
    implementations omit — producing output that decodes correctly in their own
    code and garbage everywhere else.

SIZE, MEASURED
--------------
A 640x480 circular blob encodes to ~607 characters (0.6 KB) for 307,200 pixels.
That is what makes JSONB storage viable and is why masks are not kept as PNGs
in object storage: no second fetch to render, and no orphaned objects to reap
when an annotation is deleted.

Worst case is a checkerboard — every pixel starts a run, so the encoding is
larger than the pixel count. Real annotation masks are contiguous blobs and
nowhere near that, but `validate_rle` caps size so a pathological or hostile
payload cannot bloat a row unbounded.
"""
from __future__ import annotations

# A mask whose encoding exceeds this is refused. Comfortably above any real
# annotation (the 640x480 blob above is 0.6 KB) while bounding what a single
# malformed or hostile request can write into a row.
MAX_RLE_CHARS = 2_000_000


class MaskError(ValueError):
    """Malformed mask payload."""


def encode_counts(counts: list[int]) -> str:
    """Run lengths -> compressed ASCII."""
    out: list[str] = []
    for i, value in enumerate(counts):
        x = value
        if i > 2:
            x -= counts[i - 2]
        more = True
        while more:
            chunk = x & 0x1F
            x >>= 5
            # Keep going while the remaining bits are more than sign extension.
            more = (x != -1) if (chunk & 0x10) else (x != 0)
            if more:
                chunk |= 0x20
            out.append(chr(chunk + 48))
    return "".join(out)


def decode_counts(s: str) -> list[int]:
    """Compressed ASCII -> run lengths. Inverse of :func:`encode_counts`."""
    counts: list[int] = []
    i, n = 0, len(s)
    while i < n:
        x, k, more = 0, 0, True
        while more:
            if i >= n:
                raise MaskError("Truncated RLE string")
            c = ord(s[i]) - 48
            x |= (c & 0x1F) << (5 * k)
            more = bool(c & 0x20)
            i += 1
            k += 1
            if not more and (c & 0x10):
                x |= -1 << (5 * k)       # sign-extend a negative delta
        if len(counts) > 2:
            x += counts[len(counts) - 2]
        counts.append(x)
    return counts


def rle_area(counts: list[int]) -> int:
    """Foreground pixel count. Odd-indexed runs are foreground (runs start with
    background), so this is just their sum."""
    return sum(counts[1::2])


def rle_bbox(counts: list[int], height: int, width: int) -> tuple[int, int, int, int]:
    """Tight bounding box of the foreground, as absolute (x, y, w, h).

    COCO annotations carry `bbox` alongside `segmentation`, and consumers use it
    for cropping and for mAP matching, so it must be the real extent rather than
    the whole frame.
    """
    x0, y0, x1, y1 = width, height, -1, -1
    pos, val = 0, 0
    for run in counts:
        if val and run:
            # Column-major: index = col * height + row.
            start, end = pos, pos + run - 1
            c0, r0 = divmod(start, height)
            c1, r1 = divmod(end, height)
            x0, x1 = min(x0, c0), max(x1, c1)
            if c0 == c1:
                y0, y1 = min(y0, r0), max(y1, r1)
            else:
                # The run wraps at least one column boundary, so it spans every
                # row somewhere within it.
                y0, y1 = 0, height - 1
        pos += run
        val ^= 1
    if x1 < 0:
        return (0, 0, 0, 0)
    return (x0, y0, x1 - x0 + 1, y1 - y0 + 1)


def validate_rle(mask: dict, *, width: int | None = None, height: int | None = None) -> dict:
    """Check a client-supplied mask and return it normalized.

    Shape is COCO's: ``{"size": [height, width], "counts": "<ascii>"}``.

    Validated rather than trusted because this is written straight to the
    database and read back by exporters: a counts string whose runs do not sum
    to height*width produces a mask that silently misaligns in every downstream
    consumer, and the symptom would be "the exported masks are subtly wrong",
    which is close to undebuggable after the fact.
    """
    if not isinstance(mask, dict):
        raise MaskError("mask must be an object")

    size = mask.get("size")
    counts = mask.get("counts")
    if not isinstance(size, (list, tuple)) or len(size) != 2:
        raise MaskError("mask.size must be [height, width]")
    if not isinstance(counts, str) or not counts:
        raise MaskError("mask.counts must be a non-empty RLE string")
    if len(counts) > MAX_RLE_CHARS:
        raise MaskError(f"mask.counts exceeds {MAX_RLE_CHARS} characters")

    try:
        h, w = int(size[0]), int(size[1])
    except (TypeError, ValueError) as exc:
        raise MaskError("mask.size entries must be integers") from exc
    if h <= 0 or w <= 0:
        raise MaskError("mask.size entries must be positive")

    # The mask must describe the image it is attached to, or coordinates mean
    # nothing. Checked here because the API layer knows the asset dimensions.
    if height is not None and width is not None and (h != height or w != width):
        raise MaskError(
            f"mask.size {h}x{w} does not match the image ({height}x{width})"
        )

    decoded = decode_counts(counts)
    if any(c < 0 for c in decoded):
        raise MaskError("RLE decoded to a negative run length")
    total = sum(decoded)
    if total != h * w:
        raise MaskError(
            f"RLE runs sum to {total} but the mask is {h}x{w} = {h * w} pixels"
        )

    return {"size": [h, w], "counts": counts}

File: frontend/Frontend/src/app/pages/annotate/maskCodec.ts
/**
 * COCO run-length encoding for segmentation masks — browser side.
 *
 * MUST STAY BYTE-IDENTICAL to data-intelligence-system/labeling/mask_codec.py.
 * The browser encodes what the user paints; Python decodes it for export. A
 * divergence between the two does not throw — it produces masks that decode to
 * the wrong pixels, and the only symptom is "the exported segmentations look
 * slightly off" long after the data was written. The two implementations are
 * cross-checked against shared vectors rather than trusted to agree.
 *
 * Format notes (see the Python module for the full description):
 *  - runs are counted COLUMN-MAJOR, starting with a background run that may be
 *    length 0
 *  - from the third run onward the delta against run i-2 is encoded
 *  - each value is split into 6-bit chunks, low bits first, 0x20 = "more
 *    follows", each chunk offset by 48 into printable ASCII
 *
 * JAVASCRIPT-SPECIFIC HAZARDS, both handled below:
 *  1. Bitwise operators coerce to int32. Run lengths are bounded by width x
 *     height, so a 4K image (~8.3M) needs 23 bits and is safe; deltas are
 *     smaller still.
 *  2. Shift counts are taken mod 32, so `-1 << 35` silently becomes `-1 << 3`.
 *     Python has no such wrap. The decoder therefore only sign-extends while
 *     the shift is in range — reachable in principle, and a wrong answer rather
 *     than an error if ignored.
 */

export interface RleMask {
  /** [height, width] in absolute pixels. */
  size: [number, number];
  /** Compressed run lengths. */
  counts: string;
}

/** Run lengths -> compressed ASCII. */
export function encodeCounts(counts: number[]): string {
  const out: string[] = [];
  for (let i = 0; i < counts.length; i++) {
    let x = counts[i];
    // Adjacent columns of a real mask are strongly correlated, so from the
    // third run on the delta is small and costs fewer chunks. This is where
    // the compression comes from.
    if (i > 2) x -= counts[i - 2];
    let more = true;
    while (more) {
      let chunk = x & 0x1f;
      // Arithmetic shift (>>), not logical (>>>): negative deltas must keep
      // their sign, matching Python's >>.
      x >>= 5;
      more = (chunk & 0x10) !== 0 ? x !== -1 : x !== 0;
      if (more) chunk |= 0x20;
      out.push(String.fromCharCode(chunk + 48));
    }
  }
  return out.join("");
}

/** Compressed ASCII -> run lengths. Inverse of {@link encodeCounts}. */
export function decodeCounts(s: string): number[] {
  const counts: number[] = [];
  let i = 0;
  const n = s.length;
  while (i < n) {
    let x = 0;
    let k = 0;
    let more = true;
    while (more) {
      if (i >= n) throw new Error("Truncated RLE string");
      const c = s.charCodeAt(i) - 48;
      x |= (c & 0x1f) << (5 * k);
      more = (c & 0x20) !== 0;
      i += 1;
      k += 1;
      // Sign-extend a negative delta once the final chunk is consumed. Guarded
      // because JS takes shift counts mod 32 — without the check, 5*k >= 32
      // would wrap and corrupt the value instead of failing.
      if (!more && (c & 0x10) !== 0 && 5 * k < 32) {
        x |= -1 << (5 * k);
      }
    }
    if (counts.length > 2) x += counts[counts.length - 2];
    counts.push(x);
  }
  return counts;
}

/**
 * Binary mask (row-major, 1 byte per pixel, non-zero = foreground) -> RLE.
 *
 * The input is row-major because that is what `CanvasRenderingContext2D`
 * produces; the walk below reads it column-major, which is what COCO requires.
 * Getting that backwards is the single most likely defect here and is pinned by
 * a test.
 */
export function maskToRle(data: Uint8Array | Uint8ClampedArray, height: number, width: number): RleMask {
  const counts: number[] = [];
  let prev = 0;
  let run = 0;
  for (let c = 0; c < width; c++) {
    for (let r = 0; r < height; r++) {
      const v = data[r * width + c] ? 1 : 0;
      if (v === prev) {
        run += 1;
      } else {
        counts.push(run);
        run = 1;
        prev = v;
      }
    }
  }
  counts.push(run);
  return { size: [height, width], counts: encodeCounts(counts) };
}

/** RLE -> binary mask, row-major, 1 byte per pixel. Inverse of {@link maskToRle}. */
export function rleToMask(rle: RleMask): Uint8Array {
  const [height, width] = rle.size;
  const out = new Uint8Array(height * width);
  const counts = decodeCounts(rle.counts);
  let idx = 0;
  let val = 0;
  for (const run of counts) {
    if (val) {
      for (let k = 0; k < run; k++) {
        const pos = idx + k;
        if (pos >= height * width) break;
        // Column-major position -> row-major offset.
        const r = pos % height;
        const c = (pos / height) | 0;
        out[r * width + c] = 1;
      }
    }
    idx += run;
    val ^= 1;
  }
  return out;
}

/** Foreground pixel count, without materialising the mask. */
export function rleArea(rle: RleMask): number {
  const counts = decodeCounts(rle.counts);
  let area = 0;
  for (let i = 1; i < counts.length; i += 2) area += counts[i];
  return area;
}

/** True when the mask has no foreground — used to discard empty strokes. */
export function rleIsEmpty(rle: RleMask): boolean {
  return rleArea(rle) === 0;
}
